### PR TITLE
8261125: Move VM_Operation to vmOperation.hpp

### PR DESCRIPTION
**Pull request by iklam**
vmOperations.hpp declares the VM_Operation class, as well as a hodge podge of subclasses such as VM_ForceSafepoint, VM_DeoptimizeFrame.

Out of the 1000 hotspot .o files, about 680 include vmOperations.hpp (mostly transitively). In most cases, they just need to use the VM_Operation class.

So we should move VM_Operation to its own header: vmOperation.hpp (no "s").

After the refactoring, vmOperations.hpp is included only 64 times. The inclusion count of threadSMR.hpp is also reduced from 687 to 99. HotSpot build time is improved by about 0.4%.

Tested with mach5: tier1, builds-tier2, builds-tier3, builds-tier4 and builds-tier5. Also locally: aarch64, arm, ppc64, s390, x86, and zero.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8261125](https://bugs.openjdk.java.net/browse/JDK-8261125): Move VM_Operation to vmOperation.hpp

### Reviewers
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to f4637f7050d48acd516d001dc612721be8b40f40
 * [Thomas Stuefe](https://openjdk.java.net/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to f4637f7050d48acd516d001dc612721be8b40f40

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2398/head:pull/2398`
`$ git checkout pull/2398`

### CODE REVIEW

**Review by coleenp — APPROVED** [review]
Ok!

**Review by tstuefe — APPROVED** [review]
Hi Ioi,

I like all these include cleanups! How do you find these, do you analyze the include tree?

I think vmOperation vs vmOperations could be confusing. But have no immediate better idea. If others are fine with it, I am too.

Can you please enable GA so we see that our weirder platforms build?

Otherwise, looks good.

..Thomas

**Comment by dcubed-ojdk** [comment]
A drive-by comment...

The class is named VM_Operation, but the file is named vmOperation.hpp.
The missing '_' in the file name is a bit of a disconnect, but there are worse
in the code base...

File: src/hotspot/share/runtime/vmOperation.hpp [hunk 1]
```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright (c) 1997, 2021, Oracle and/or its affiliates. All rights reserved.
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/vmOperation.hpp:2]: New file so single copyright year.

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
Hi Ioi,

The distinction between vmOperation versus vmOperations is far too subtle. Perhaps as Dan implied vm_Operation.hpp or VM_Operation.hpp (though that breaks normal - odd - naming convention).

I assume most files that include vmOperation.hpp are those that define VM_Operation subclasses?

Thanks,
David

**Comment by iklam** [comment]
> I like all these include cleanups! How do you find these, do you analyze the include tree?

I have a few hand-rolled tools for identifying what header files are easy/beneficial to refactor. See

https://github.com/iklam/tools/tree/main/headers

> I think vmOperation vs vmOperations could be confusing. But have no immediate better idea. If others are fine with it, I am too.
> 
> Can you please enable GA so we see that our weirder platforms build?

I think I finally re-enabled the GitHub actions properly for my PRs. Let's see if they get run for my next push to this PR :-)

**Comment by iklam** [comment]
> Hi Ioi,
> 
> The distinction between vmOperation versus vmOperations is far too subtle. Perhaps as Dan implied vm_Operation.hpp or VM_Operation.hpp (though that breaks normal - odd - naming convention).

How about this:

- runtime/vmOperation.hpp   --- (new file) this is the file that declares VM_Operation
- runtime/commonVMOperations.hpp -- (renamed from vmOperations.hpp) these are the VM_Operation subclasses that no one cares to organize properly :-)

this will be kinda consistent with these existing files:

- gc/g1/g1VMOperations.hpp
- gc/g1/g1VMOperations.cpp
- gc/shenandoah/shenandoahVMOperations.cpp
- gc/shenandoah/shenandoahVMOperations.hpp
- gc/shared/gcVMOperations.cpp
- gc/shared/gcVMOperations.hpp
- gc/parallel/psVMOperations.cpp
- gc/parallel/psVMOperations.hpp

(I should also add a new vmOperation.cpp, and rename vmOperations.cpp to commonVMOperations.cpp)

BTW, I need to refactor VM_Exit into its own file. It's used by the `JVM_LEAF` macro in interfaceSupport.inline.hpp, but I don't want to pull in all the other "common" operations in there.  I am thinking of calling it vmExit.hpp (since exitVMOperation.hpp doesn't really look good).

> I assume most files that include vmOperation.hpp are those that define VM_Operation subclasses?

Yes, but there are also files that use the `VM_Operation::VMOp_Type` type, notably safepoint.hpp.

**Comment by coleenp** [comment]
I'm fine with leaving vmOperations.hpp and vmOperation.hpp.  It's not a big deal.  commonVMOperations.hpp - too much noise!
I agree with David.  interfaceSupport.inline.hpp imports a lot of things so importing vmOperations.hpp is not a big deal.  vmOperations.hpp imports #include "runtime/threadSMR.hpp" otherwise it has all the same imports as interfaceSupport.inline.hpp anyway.
All these files are going to increase compilation time too.

**Comment by iklam** [comment]
> I'm fine with leaving vmOperations.hpp and vmOperation.hpp. It's not a big deal. commonVMOperations.hpp - too much noise!
> I agree with David. interfaceSupport.inline.hpp imports a lot of things so importing vmOperations.hpp is not a big deal. vmOperations.hpp imports #include "runtime/threadSMR.hpp" otherwise it has all the same imports as interfaceSupport.inline.hpp anyway.
> All these files are going to increase compilation time too.
> I stand by my check mark above!

OK, since no one is fond of further splitting the headers, and no one has vetoed the vmOperation.hpp name, I'll keep everything as originally proposed. Will do more testing and integrate.

**Comment by iklam** [comment]
BTW, I found a few existing singular/plural pairs of hpp files. Admittedly I created ciSymbols.hpp recently, but the other 3 pairs have been there for quite some time.

share/ci/ciSymbol.hpp
share/ci/ciSymbols.hpp

share/interpreter/bytecode.hpp
share/interpreter/bytecodes.hpp

share/jfr/recorder/service/jfrEvent.hpp
share/jfr/jfrEvents.hpp

share/jfr/recorder/checkpoint/types/jfrType.hpp
share/jfr/utilities/jfrTypes.hpp

**Comment by iklam** [comment]
Thanks @dcubed-ojdk @coleenp @tstuefe @dholmes-ora for the review.
/integrate

**Pull request closed without merging**